### PR TITLE
[Explorer] Networks change menu

### LINKED ISSUE #160: Quick change networks menu
image

* We would allow the user to change quickly from one network to the other using the header.
* We would respect the relative path when we do it
* This means that if I'm in rinkeby viewing one order ("/rinkeby/orders/1234"), and I navigate to  xDAI, it will redirect to the same page of the new network ("/xdai/orders/1234"). This way, if the order was not found because we are in the wrong network, we can easily find it in the right one!

### PR DESCRIPTION
**Pull request by matextrem**
# Summary

Closes #160 

Added a networks menu on header in order to change between networks quickly

Screen Shot 2021-12-09 at 11 13 18

# To Test

1.  Open any page in Explorer.
* Click on network name next to header.
* A selector will appear with the different available networks to change.
* Clicking on any of the networks will redirect to the explorer network domain.

### CODE REVIEW

File: src/components/NetworkSelector/NetworkSelector.styled.ts
Comment (alongoni) [inline src/components/NetworkSelector/NetworkSelector.styled.ts:96]: Maybe we can think another color for Ethereum.
image

**Comment by elena-zh** [comment]
I'd change networks order to 
 - Ethereum (without 'mainnet' as we do in CowSwap 
image )
 - xDAI
 - Rinkeby

**Comment by elena-zh** [comment]
It would be nice to close the networks dropdown when click outside of its area, IMO. See the video: https://watch.screencastify.com/v/S5NVJUWEXBtSDz52jLXw

**Comment by elena-zh** [comment]
Orders not found in < network > is displayed for a while before orders are loaded when switch networks from a user details page. See the video: I switch to Rinkeby when I have orders, but before orders start loading, I see message 'Orders not found in Rinkeby'
https://watch.screencastify.com/v/PzdKGgtgkzZrVUDf10AS

It would be nice to show a loader right after network switching event.

**Comment by elena-zh** [comment]
'Back home' button from 404 page always navigated back to the home page in the Mainnet: 
https://watch.screencastify.com/v/ifcvWMAE7V9wiJttGb5T

**Comment by elena-zh** [comment]
The question about navigation from an order details/transaction details pages is still valid
https://github.com/gnosis/gp-ui/issues/160#issuecomment-988923833

**Note**: transaction details page is currently under implementation

File: src/components/NetworkSelector/NetworkSelector.styled.ts
Comment (alongoni) [reply src/components/NetworkSelector/NetworkSelector.styled.ts:96]: #62688f
https://ethereum.org/en/assets/

File: src/components/NetworkSelector/index.tsx
Comment (anxolin) [inline src/components/NetworkSelector/index.tsx:22]: Why not just Ethereum (like we do in CowSwap)

File: src/components/NetworkSelector/NetworkSelector.styled.ts
Comment (anxolin) [inline src/components/NetworkSelector/NetworkSelector.styled.ts:93]: why rinkeby doesn't have a color? I would use the same as Metamask (yellow)

image

File: test/utils/url.spec.ts
Comment (anxolin) [inline test/utils/url.spec.ts:3]: Are these test related to this PR?

**Review by anxolin — APPROVED** [review]

File: test/utils/url.spec.ts
Comment (matextrem) [reply test/utils/url.spec.ts:3]: Yes, they are.

File: src/apps/explorer/layout/Header.tsx
Comment (alongoni) [inline src/apps/explorer/layout/Header.tsx:44]: The color for Rinkeby could be: #f6c343 and text in a dark color
image

Based on MM (as Anxo suggested)
image

**Review by alongoni — APPROVED** [review]
Looks great!
(only one comment regarding Rinkeby color)

**Comment by elena-zh** [comment]
@matextrem , FYI: 

> As we discussed during our sync, let's redirect the user to the landing page when switching networks.
> 
> That is, for details pages. For list pages, it should remain in the same place
> 
> Example:
> 
> 1. User is on order details for `mainnet` (https://gnosis-protocol.io/orders/0x.../) and selects `xdai` => go to `https://gnosis-protocol.io/xdai/`
> 2. User is on tx details for `rinkeby` (https://gnosis-protocol.io/rinkeby/tx/0x..../) and selects `mainnet` => go to `https://gnosis-protocol.io/`
> 3. User on address page on `xdai` (https://gnosis-protocol.io/xdai/address/0x...) and selects `rinkeby` => go to `https://gnosis-protocol.io/rinkeby/address/0x...`


For more details please refer here: https://github.com/gnosis/gp-ui/issues/160#issuecomment-992768181

File: src/api/operator/operatorApi.ts
Comment (alfetopito) [inline src/api/operator/operatorApi.ts:260]: Ops

```suggestion
```

Also, the same info is already logged above

File: src/components/NetworkSelector/index.tsx
Comment (alfetopito) [inline src/components/NetworkSelector/index.tsx:58]: Why the `/mainnet` prefix?

File: src/components/NetworkSelector/index.tsx
Comment (alfetopito) [inline src/components/NetworkSelector/index.tsx:51]: Can I make one more request?

Dismiss the modal using `esc`

File: test/utils/url.spec.ts
Comment (alfetopito) [inline test/utils/url.spec.ts:16]: `mainnet` links have no prefix

**Review by alfetopito — CHANGES_REQUESTED** [review]
Looks amazing and works 100%, love it!

Left a few small comments.

Besides the rinkeby color, the thing that needs changing is the `/mainnet` redirect.

Regarding the behaviour I commented on #160 and Elena linked here, it can be addressed on another PR

File: test/utils/url.spec.ts
Comment (matextrem) [reply test/utils/url.spec.ts:16]: I used it in order to keep the same logic for the regex `/network/... ` since `/mainnet` is redirected to `/`

File: src/components/NetworkSelector/index.tsx
Comment (alfetopito) [inline src/components/NetworkSelector/index.tsx:69]: Can you move this to a const?

File: src/apps/explorer/pages/NotFound.tsx
Comment (alfetopito) [inline src/apps/explorer/pages/NotFound.tsx:61]: Lol, just noticed the component name.

Could you give one more meaningful?

File: src/components/NetworkSelector/index.tsx
Comment (alfetopito) [inline src/components/NetworkSelector/index.tsx:56]:
```suggestion
      } else if (selectContainer.current && !selectContainer.current.contains(e.target as HTMLElement)) {
        setOpen(false)
      }
```

File: test/utils/url.spec.ts
Comment (alfetopito) [reply test/utils/url.spec.ts:16]: I rather have the extra logic inside `replaceURL` function instead of being on the caller to preprend `/mainnet`

**Comment by elena-zh** [comment]
@matextrem , Looks and works like a charm to me!

However, a tiny question: when I search for an order from a different network, I'm automatically redirected to another network when an order is loaded: https://watch.screencastify.com/v/IsUCHwo38NzN1Gq0fTnL

Should we leave it as it is? Or we should redirect a user to search page? 
Or we should show a message that the order is not found in the required network, but found in another one, as we do when we search for a user details. 
image

@alfetopito , WDYT?

**Review by elena-zh — APPROVED** [review]

**Comment by alfetopito** [comment]
> @matextrem , Looks and works like a charm to me!
> 
> However, a tiny question: when I search for an order from a different network, I'm automatically redirected to another network when an order is loaded: https://watch.screencastify.com/v/IsUCHwo38NzN1Gq0fTnL
> 
> Should we leave it as it is? Or we should redirect a user to search page? Or we should show a message that the order is not found in the required network, but found in another one, as we do when we search for a user details. image
> 
> @alfetopito , WDYT?

I like the current behaviour for orders, because it's not likely to have an order with same ID in different networks.
So we either find it - and display - for the right network or not at all.
Different than when looking for addresses where it can exist in multiple networks.

File: src/utils/url.ts
Comment (alfetopito) [inline src/utils/url.ts:22]: Please don't be angry with me hahaha

Just one last request.

Can you exit early when mainnet? It doesn't need the other parts and will look cleaner.
Something like:

```
  if (currentNetwork === Network.Mainnet) {
    return url.replace(/^/, `/${strReplace}`)
  }
  ... the rest here
```

After that you can merge it directly, you already got the approvals ;)

**Review by alfetopito — APPROVED** [review]

**Pull request merged**
